tawazi/_dag/constructor.py

```diff
@@ -0,0 +1,121 @@
+import inspect
+import warnings
+from typing import Any, Callable, Dict, List, Tuple, Union
+
+from tawazi._helpers import StrictDict
+from tawazi.consts import RVDAG, P
+from tawazi.node import ArgExecNode, ExecNode, ReturnUXNsType, UsageExecNode, node, wrap_in_uxns
+from tawazi.node.node import make_axn_id
+
+from .dag import DAG, AsyncDAG
+
+
+def get_args_and_default_args(func: Callable[..., Any]) -> Tuple[List[str], Dict[str, Any]]:
+    """Retrieves the arguments names and the default arguments of a function.
+
+    Args:
+        func: the target function
+
+    Returns:
+        A Tuple containing a List of argument names of non default arguments,
+         and the mapping between the arguments and their default value for default arguments
+
+    >>> def f(a1, a2, *args, d1=123, d2=None): pass
+    >>> get_args_and_default_args(f)
+    (['a1', 'a2', 'args'], {'d1': 123, 'd2': None})
+    """
+    signature = inspect.signature(func)
+    args = []
+    default_args = {}
+    for k, v in signature.parameters.items():
+        if v.default is not inspect.Parameter.empty:
+            default_args[k] = v.default
+        else:
+            args.append(k)
+
+    return args, default_args
+
+
+def __make_dag(
```

Comment:
why the trailing double _ in the fn name ? Wouldn't one be enough ?

Reply:
done